[WIP]Use do_settings_change in organisation settings.

shubhamdhama opened the pull request:
**Testing Plan:** <!-- How have you tested? -->
Manula Testing.

This is a WIP in PR currently, so before proceeding to implement do_settings_change in organization settings section, I want to make sure that the current two commits makes sense now or not. (Don't merge them now, as I wanna test it further while implementing them in org settings page .)
@timabbott what's your view on this?

timabbott commented:
Yeah, those draft changes look great to me!

shubhamdhama commented:
@timabbott I've added some more commits, these are the changes needed to further implement what discussed here at https://chat.zulip.org/#narrow/stream/101-design/subject/New.20save.20feedback.20system/near/538317.
I'll add commit for new org-settings saving mechanism soon.

timabbott commented:
The approach for the commits so far looks good to me.  Haven't done manual testing, so I'm not merging anything yet.

shubhamdhama commented:
(I haven't updated the tests so tests will fail.) 
For failed changes, showing status doesn't look great in save button and so I've kept just one status element for failed changes and made "Saving" button to reset to "Save" in error callback. 

I've also restructured the organization settings page into more sensible subsection(with no change in UI).
(Currently, I'm updating node and Casper tests).
@timabbott FYI.

(I hope commits are properly split, I've tested each commit and they work properly individually.
Another thing I want to know, currently I'm updating tests for each commit but is this fine to update tests for all commit in the last commit or separate commit).

shubhamdhama commented:
So, while writing node tests I'm facing errors, to solve them I first need to make sure that are we okay my current approach.
In my current approach, I'm getting the subsection properties names from their `id`'s by some minor string operations. 
If there's better approach please give me lead about that.

timabbott commented:
@shubhamdhama I'll take a look at this.

timabbott commented:
I merged the first 5 preparatory commits, since they looked good and don't make any functional changes.  

Playing with the end of the branch, the "Discard changes" buttons should hide when you click "Save changes"; otherwise the final end state seems to work pretty well.  

I also noticed that we seem to have a bug in the real-time sync if you have 2 windows open (if you click "save changes" in one window, the other's checkbox state doesn't update), but that's probably not a regression here, just a follow-up issue.

As a sidenote, for next steps, we should start thinking about using a handlebars template for elements like the the `organization-submission`  block, so that we can deduplicate the code (both for those blocks and eventually also for our checkbox elements...).  Probably not worth thinking about until this is merged though.

timabbott commented:
BTW, the first 5 commits were exactly the sort of thing I was looking for in terms of commit structure :)

timabbott commented:
For "org settings: Make each subsection independent for saving changes.", I think you could without too much trouble split the code structure changes out into a preparatory commit that doesn't include the CSS/template changes.

For the node tests, your high-level approach looks great.  I think @showell wrote that test file, so if you're having trouble debugging, ask him for help.

timabbott commented on static/js/settings_org.js:
```diff
@@ -394,106 +418,159 @@ function _set_up() {
+            if (typeof page_params[property_name] === 'boolean') {
+                // Trigger a click event because of dependent sub-settings
+                if (element.prop('checked') !== page_params[property_name]) {
+                    element.click();
```
I think for flipping a checkbox back to its default state, it'd be better to use something lower-level that triggering a click event (e.g. just setting the prop).

shubhamdhama replied:
Yeah, I've first used the `prop` but for cases like `Users can edit their messages` we have to also toggle dependent field like `Message edit limit in minutes`. So, in that case `click()` event holds better. But yes, for just one case it's not good to trigger `click()` event on all other fields. Any better suggestion.

shubhamdhama replied:
And I think there is no better way to determine if any field is dependent on another currently.
@timabbott so should leave as it is?

timabbott replied:
I see.  If that's the reasoning, just leave a comment like this;
```
We trigger a click event, rather than just changing the prop here, because that automatically handles the logic around enabling/disabling dependent sub-settings.
```

timabbott commented on static/js/settings_org.js:
```diff
@@ -394,106 +418,159 @@ function _set_up() {
                 }
+            } else if (typeof page_params[property_name] === 'string' || typeof page_params[property_name] === 'number') {
+                element.val(page_params[property_name]);
+            } else if (typeof page_params[property_name] === "undefined") {
```
Can you add a comment explaining what this case is for?

timabbott commented on static/js/settings_org.js:
```diff
@@ -394,106 +418,159 @@ function _set_up() {
+        } else if (current_val === undefined) {
+            current_val = property_value_element_refers(property_name).toString();
+            changed_val = elem.val().trim();
         }
```
I would have an `else` case that throws a blueslip error here.

timabbott commented on static/js/settings_org.js:
```diff
@@ -394,106 +418,159 @@ function _set_up() {
-            }
+    function check_property_changed(elem) {
+        elem = $(elem);
+        var property_name = elem.attr('id').replace("id_", "").split('-').join('_');
```
This conversion might be written more clearly as a function that has a few node unit tests for specific cases that we use, since it's complicated enough that one might wonder if it's correct in every case.

shubhamdhama commented:
I've updated the PR for last three reviews but still doubt about first remains https://github.com/zulip/zulip/pull/8650#discussion_r175105634.
This time I've also updated current Casper Tests which are passing currently but I think it would be great if more Casper tests are added for the changes.
I'll forward to node tests if current changes look right.

timabbott commented on static/js/settings_org.js:
```diff
@@ -418,106 +418,177 @@ function _set_up() {
+            // Check whether the id refers to a property whose name we can't
+            // extract from element's id.
+            current_val = property_value_element_refers(property_name).toString();
+            changed_val = elem.val().trim();
```
I think this code path will crash if `current_val` is undefined (and actually the `toString` will crash if `property_value_element_refers` is undefined); seems probably buggy.

shubhamdhama commented:
Okay, I'm taking a deep look at all possible bugs. BTW, for now, I've updated tests (Though I'm not personally happy with the tests for submit_settings). But I'll update them here or as a followup.

shubhamdhama commented:
@timabbott can you give this another round of review, I've added a "data-status" attribute to track the last status of the element, which helps in "Casper test" and for further tracking the status of saving.
If this gets finalized I'll extract `if (subsection === 'msg_editing') {` into a separate function to make this clean when we use this function for org-permissions, and also will add some bluesip errors for subsection validity(if this is important here).

timabbott commented:
@shubhamdhama this looks pretty reasonable; I expect to merge it tomorrow after some manual testing.

timabbott commented on static/js/settings_org.js:
```diff
@@ -418,104 +418,177 @@ function _set_up() {
+    function check_property_changed(elem) {
+        elem = $(elem);
+        var property_name = exports.extract_property_name(elem);
+        var current_val = page_params[property_name];
```
This seems like a good place to assert that this exists.

shubhamdhama replied:
For settings like `realm_message_edit_limit_in_minutes`(which doesn't exist in page_params) we have to do the conversion that's why I've checked it later in else statement after checking if this refers to any valid property_name or not.

timabbott replied:
OK, makes sense.

timabbott commented:
Merged, with these small changes squashed into the commits in question:
```
diff --git a/static/js/settings_org.js b/static/js/settings_org.js
index aa021f4..ff5aa7f 100644
--- a/static/js/settings_org.js
+++ b/static/js/settings_org.js
@@ -492,7 +492,9 @@ function _set_up() {
             elem = $(elem);
             var property_name = exports.extract_property_name(elem);
             if (typeof page_params[property_name] === 'boolean') {
-                // Trigger a click event because of dependent sub-settings
+                // We trigger a click event rather than just changing
+                // the prop, because that handles dependent
+                // sub-settings correctly.
                 if (elem.prop('checked') !== page_params[property_name]) {
                     elem.click();
                 }
@@ -562,9 +564,9 @@ function _set_up() {
             if (!new_allow_message_editing) {
                 if ((parseInt(compose_textarea_edit_limit_minutes, 10).toString() !==
                      compose_textarea_edit_limit_minutes) ||
-                    compose_textarea_edit_limit_minutes < 0) {
-                // Realm.DEFAULT_MESSAGE_CONTENT_EDIT_LIMIT_SECONDS / 60
-                compose_textarea_edit_limit_minutes = 10;
+                        compose_textarea_edit_limit_minutes < 0) {
+                    // Realm.DEFAULT_MESSAGE_CONTENT_EDIT_LIMIT_SECONDS / 60
+                    compose_textarea_edit_limit_minutes = 10;
                 }
             }
 
```
Thanks @shubhamdhama!  This is a big improvement to that screen.

I think the natural follow-up projects for this are:
* Fixing the real-time sync for having 2 windows open and editing these settings in one of them.  We'll need to do something similar to what we have in "notification settings" and "display settings" for this.
* Doing the same treatment to "organization permissions"
* Starting to work on de-duplicating the handlebars template duplicated code.

I think the first two would be things we could merge for the upcoming release, so perhaps it's best to focus on those.
